src/receiver/google_cloud/mod.rs

```diff
@@ -0,0 +1,162 @@
+            update_freq,
+            network,
+            bucket,
+        };
+
+        let (worker_data_sender, worker_data_receiver) = watch::channel(worker_data.clone());
+
+        let mut worker = GoogleCloudBlockWorker::new(
+            worker_data,
+            blocks_producer,
+            error_sender,
+            command_receiver,
+            worker_data_sender,
+        )?;
+
+        let worker_join_handle = tokio::spawn(async move {
+            worker.start_loop().await;
+        });
+
+        Ok(Self {
+            blocks_consumer,
+            error_receiver,
+            command_sender,
+            worker_data_receiver,
+            worker_join_handle: Some(worker_join_handle),
+        })
+    }
+
+    pub async fn set_worker_data(
+        &self,
+        worker_data: GoogleCloudBlockWorkerData,
+    ) -> Result<(), GoogleCloudBlockReceiverError> {
+        self.command_sender
+            .send(GoogleCloudBlockWorkerCommand::SetWorkerData(worker_data))
+            .await
+            .map_err(GoogleCloudBlockReceiverError::CommandError)
+    }
+
+    pub async fn get_worker_data(
+        &self,
+    ) -> Result<GoogleCloudBlockWorkerData, GoogleCloudBlockReceiverError> {
+        self.command_sender
+            .send(GoogleCloudBlockWorkerCommand::GetWorkerData)
+            .await
+            .map_err(GoogleCloudBlockReceiverError::CommandError)?;
+        Ok(self.worker_data_receiver.borrow().clone())
+    }
+}
+
+#[async_trait]
+impl BlockReceiver for GoogleCloudBlockReceiver {
+    async fn recv_block(&mut self) -> Result<Option<PrecomputedBlock>, anyhow::Error> {
+        tokio::select! {
+            block = self.blocks_consumer.pop() => {
+                return Ok(block);
+            },
+            error = self.error_receiver.changed() => {
+                match error {
+                    Ok(_) => {
+                        let error = self.error_receiver.borrow().clone();
+                        return Err(error.expect("error channel only changes when an error is present").into());
+                    },
+                    Err(receiver_error) => return Err(receiver_error.into()),
+                }
+            }
+        }
+    }
+}
+
+impl Drop for GoogleCloudBlockReceiver {
+    fn drop(&mut self) {
+        let command_sender = self.command_sender.clone();
+        let worker_join_handle = self.worker_join_handle.take();
+        let temp_block_dir = self.worker_data_receiver.borrow().temp_blocks_dir.clone();
+        tokio::spawn(async move {
+            command_sender
+                .clone()
+                .blocking_send(GoogleCloudBlockWorkerCommand::Shutdown)
+                .expect("shutdown command sends correctly");
+
+            if let Some(join_handle) = worker_join_handle {
+                join_handle.await.expect("worker fininshes successfully");
+            }
+        });
+        std::fs::remove_dir_all(temp_block_dir).expect("block dir exists");
```

Comment:
This is okay because it's just a temporary directory after all

Reply:
we can remove the one in the command handler just fine, it doesn't do anything